mesa/mesaPROTON_OC.py

```diff
@@ -1166,6 +1176,136 @@ def let_migrants_in(self):
+            for agent in extra.weighted_n_of(target_n_of_arrest, criminals, lambda x: x.arrest_weight, self.rng):
+                agent.get_caught()
+
+    def calculate_oc_status(self, co_offenders):
+        """
+        This procedure modify in-place the arrest_weigh attribute of the Person objects passed to co_offenders
+        :param co_offenders: list, of Person object
+        :return: None
+        """
+        for agent in co_offenders:
+            agent.arrest_weight = agent.calculate_oc_member_position()
+        min_score = np.min([agent.arrest_weight for agent in co_offenders])
+        divide_score = np.mean([agent.arrest_weight - min_score for agent in co_offenders])
+        for agent in co_offenders:
+            if divide_score > 0:
+                agent.arrest_weight = (agent.arrest_weight - min_score) / divide_score
+            else:
+                agent.arrest_weight = 1
+
+
+    def commit_crime(self, co_offenders):
+        """
+        This procedure modify in-place the num_crimes_committed,num_crimes_committed_this_tick, co_off_flag and num_co_offenses
+        attributes of the Person objects passed to co_offenders
+        :param co_offenders: list, of Person object
+        :return: None
+        """
+        for co_offender in co_offenders:
+            co_offender.num_crimes_committed += 1
+            co_offender.num_crimes_committed_this_tick += 1
+            other_co_offenders = [agent for agent in co_offenders if agent != co_offender]
+            for agent in other_co_offenders:
+                if agent not in co_offender.neighbors.get("criminal"):
+                    co_offender.addCriminalLink(agent)
+                    co_offender.co_off_flag[agent] = 0
+        for co_offender in co_offenders:
+            for co_off_key in co_offender.co_off_flag.keys():
+                co_offender.co_off_flag[co_off_key] += 1
+                if co_offender.co_off_flag[co_off_key] == 2:
+                    co_offender.num_co_offenses[co_off_key] += 1
+
+
+    def number_of_accomplices(self):
+        """
+        Pick a group size from the num. co-offenders distribution
+        and substract one to get the number of accomplices
+        :return:
+        """
+        return extra.pick_from_pair_list(self.num_co_offenders_dist, self.rng) - 1
+
+
+    def update_meta_links(self,agents):
+        """
+        This method creates a new temporal graph that is used to colculate the oc_embeddedness of an agent, the graph is stored
+        in the variable self.meta_graph
+        :param agents: list, of Person objects
+        :return: None
+        """
+        self.meta_graph = nx.Graph()
+        for agent in agents:
+            self.meta_graph.add_node(agent.unique_id)
+            for in_radius_agent in agent.agents_in_radius(1): # limit the context to the agents in the radius of interest
+                self.meta_graph.add_node(in_radius_agent.unique_id)
```

Comment:
Is it ok here not to check if the node exists? We could assume that, if it does exist, the weight has already been calculated..

Reply:
You're right but the gain is is uncertain, probably because networkx uses a linear node search which is slow and therefore compensates for the time spent calculating the weight. In my opinion, it's best to leave it as is. I leave the code below in case you want to replicate. 

**Modified function:**
```python
    def mod_update_meta_links(self,agents):
        self.meta_graph = nx.Graph()
        for agent in agents:
            self.meta_graph.add_node(agent.unique_id)
            for in_radius_agent in agent.agents_in_radius(1): 
                if in_radius_agent not in self.meta_graph.nodes:
                    self.meta_graph.add_node(in_radius_agent.unique_id)
                    w = 0
                    for net in Person.network_names:
                        if in_radius_agent in agent.neighbors.get(net):
                            if net == "criminal":
                                if in_radius_agent in agent.num_co_offenses:
                                    w += agent.num_co_offenses[in_radius_agent]
                            else:
                                w += 1
                    self.meta_graph.add_edge(agent.unique_id, in_radius_agent.unique_id, weight=1 / w)
        return self.meta_graph
```
**Test**
```python
    model.num_oc_persons = 10000
    model.setup(1000)
    import time

    start1 = time.time()
    x1 = model.update_meta_links(model.schedule.agents)
    print("update meta_links: " + str(time.time() - start1))

    start2 = time.time()
    x2 = model.mod_update_meta_links(model.schedule.agents)
    print("mod update meta_links: " + str(time.time() - start2))

    em = nx.algorithms.isomorphism.numerical_edge_match('weight', 1)
    print(nx.is_isomorphic(x1, x2))
    print(nx.is_isomorphic(x1, x2, edge_match=em))
```

**Results:**

num_oc_agents = 1000
num_agents = 1000
update meta_links: 1.736 (avg 20 runs)
mod update meta_links: 2.057 (avg 20 runs)

Reply:
Agreed!